### ISSUE TITLE
Neural Style Algorithm

### ISSUE THREAD

**Issue by schrum2** [deeplearning4j/deeplearning4j#4298]
This is a follow-up on a Gitter conversation, and not an actual problem with DL4J per se, but if anyone has ideas on how to proceed, it would be helpful.

I'm trying to implement the Neural Style Transfer algorithm in DL4J (https://arxiv.org/abs/1508.06576 and also popular via https://deepart.io/). This was mentioned in the DL4J book by Patterson and Gibson, so I naturally wonder whether anyone has tried doing this in DL4J yet. Searching the internet, there are plenty of Python implementations. I've gotten this Tensorflow implementation to work (https://github.com/anishathalye/neural-style), and have also been looking at this alternate detailed Tensorflow description: https://harishnarayanan.org/writing/artistic-style-transfer/ 

Regarding the last link above, I've been trying to follow the proposed steps while converting the Python code into Java. Here is what I have so far:
https://github.com/schrum2/MM-NEAT/blob/f40333da7e6c568c8e977d23feef235f53310151/src/main/java/edu/southwestern/util/graphics/NeuralStyleTransfer.java

Basically, I believe I've written code that correctly calculates all of the components of the loss function, but the real sticking point comes at the section under the heading "Define needed gradients and solve the optimisation problem" in the link above. At this point, I need to define the gradients for changing my combined image with respect to the loss function, and then also run an optimization algorithm, specifically L-BFGS.

DL4J supports this as an optimizer, but all of the optimizers in DL4J seem to be tailored specifically to neural networks. This algorithm doesn't change the weights of a neural network ... rather, it uses a fixed network (VGG-16 in my case, but VGG-19 also works) to calculate a loss function and change the content of an image that combines the content of one image with the style of another.

Any suggestions of how to proceed without rewriting slightly modified versions of lots of DL4J code would be appreciated ... for example, writing my own implementation of L-BFGS that generalizes beyond neural networks. One idea I had was to create a "network" that isn't really a network at all, and use its "weights" to encode the information in an image. However, I was hoping to get some feedback before diving into this, since it would likely be quite a challenge.

**Comment by AlexDBlack** [comment]
I don't believe I've seen a completed implementation using DL4J (a few people have asked about it). It would definitely be cool to get this going.

Conceptually, style transfer is not that difficult. But it's not something DL4J is designed around, so you'll need to do some non-standard stuff to get it working.

Starting with your main method:

```combination``` should not start as all zeros... from last paragraph, pg 11 of the paper:
"To generate the images that mix the content of a photograph with the style of a painting (Fig 2) *we jointly minimise the distance of a white noise image* from the content representation"
I note one of the links you have (https://harishnarayanan.org/writing/artistic-style-transfer/) starts with U(-128,128).
I'd probably generate U(0,255) and pass it through the normalizer, to start.

Score calculation looks OK at first glance.

At first glance, the stack/concat is a good idea. But keep in mind the activations won't change for the content and style images, so you might as well just calculate them once, and reuse.

Gradient calculation will be the harder thing here.
Obviously we've got the content loss (one layer) and the style loss (multiple layers).

I can think of a couple of ways to implement it.
Option 1: manually backprop through the layers. Easy for VGG, quite difficult for say inception/resnet.
Option 2: augment the network with some custom loss layers and use built-in backprop methods.

For VGG16/19, I'd probably recommend just going with option 1
Basically, you loop backwards over the layers... it'd be something along the lines of:
```
	int startLayer = net.getLayer("block2_conv2").getIndex();
	INDArray dLdANext = Nd4j.zeros(...);	//Shape: size of activatiosn of block2_conv2
	for(int i=startLayer; i>=0; i++ ){
		INDArray dLdAOut = dLdaNext.add(dLcontent_currLayer).add(dLstyle_currLayer);
		//dLcontent and dLstyle might be 0 for some layers (not sure if subsampling would be included?)

		Layer l = net.getLayer(i);
		dLdANext = l.backpropGradient(dLdAOut).getSecond();
	}

	//Once loop ends: dLdANext == dL/dCombination
```

Then, we just need to do gradient descent on the input image.
The simplest: SGD is literally just ```combination.subi(dLdANext.mul(lr))```, with some appropriate learning rate.

We can probably hack an updater to use something like Adam or Nesterov updaters. Not too hard, but more complex than SGD obviously.
The most complex would be to use DL4J's LBFGS etc optimizers. I'd need to think more about that, if you want to go down that route.

**Comment by schrum2** [comment]
Thanks so much for all of your help!

I've made some progress here
https://github.com/schrum2/MM-NEAT/blob/c300a651c3463eea88a71fb229f4a28bc7f1e288/src/main/java/edu/southwestern/util/graphics/NeuralStyleTransfer.java

The combo image starts out as white noise now. I will eventually remove the unnecessary extra activations from stacking the content, style, and combo images, but for now I just want to get something that works. As a result, I'm also fine not using LBFGS for now ... like I said, I just want to get something working.

However, after trying to follow all of your steps, there is an error.

Specifically, there is something wrong with the matrix multiplication on line 190:
INDArray product = comboFeatures.transpose().mmul(diff);
I've probably simply misunderstood the math from the Gatys et al paper (https://arxiv.org/pdf/1508.06576.pdf) ... this is supposed to be part of equation (6), as described in my comments. However, I may also be misunderstanding the ND4J operations.

However, I did want to make sure that using equations (2) and (6) from this paper is what you intended for the variables dLcontent_currLayer and dLstyle_currLayer.

I also found it off that the loss function I calculated doesn't seem to be needed, but then I realized that the reason this was used in the various Python implementations was that TensorFlow/Keras seems to automatically do all of this gradient calculation simply through specifying the variable and the tensor being optimized, which is frankly quite amazing if it is true ... any chance DL4J will eventually have that ability?

Thanks again for the help.

**Comment by AlexDBlack** [comment]
ok, so line 190: I wouldn't expect it to work...
INDArray product = comboFeatures.transpose().mmul(diff);

CNN activations in DL4J have 4 dimensions: [minibatch, depth, height, width]. NCHW format for short
the array comboFeatures is derived from this: (point(2), all, all, all) -> shape [d, h, w]
whereas comparatively, diff (and g/a) should have shape [d*h*w, d*h*w] I think
Clearly, you can't perform a matrix multiplication on these arrays: [d,h,w] x [d*h*w, d*h*w] just isn't defined mathematically

Your gram matrix itself looks correct to me (but I'm not 100% sure).

As to what that derivative *should* be: that's a good question (that unfortunately I don't have time to dig into properly, sorry). Maybe deriving dE/dF_ij from E might provide some insight, however...

**Comment by schrum2** [comment]
Thanks again for the help.

Actually, there does indeed seem to be a problem with the gram matrix calculation. It turns out the result had a size of 1 by 1. The reason for this is that I interpreted the Keras command _backend.batch_flatten_ from https://harishnarayanan.org/writing/artistic-style-transfer/ as being equivalent to the DL4J command _Nd4j.toFlattened_, but now I know this is wrong, based on the description here: https://www.tensorflow.org/api_docs/python/tf/keras/backend/batch_flatten

So, which DL4J command should I be using instead of _Nd4j.toFlattened_?

Also, regarding comboFeatures, I don't think I have a problem because the parameter that is passed into that method already has the get(point(2), all, all, all) operation performed on it.

There are probably other issues, but I definitely need to fix the gram matrix problem first.

**Comment by schrum2** [comment]
Update:
I believe that I have used the reshape method to correct process the array for calculation of the Gram Matrix. The updated code is here: https://github.com/schrum2/MM-NEAT/blob/8c7250d1c821b51e673de99bed98432bb7a33777/src/main/java/edu/southwestern/util/graphics/NeuralStyleTransfer.java

I use reshape to calculate the gram matrix:
int[] shape = x.shape();
INDArray flattened = x.reshape(shape[0],shape[1]*shape[2]);

When executing the code with the print statements that are currently in it, the result is (the arrow labels were added by me afterward):
Iteration: 0
Layer: block2_conv2
Rank: 3,Offset: 0
 Order: c Shape: [128,112,112],  stride: [37632,112,1]   <----- combo features
Rank: 2,Offset: 0
 Order: f Shape: [128,128],  stride: [1,128]             <------ gram matrix
Rank: 3,Offset: 0
 Order: c Shape: [112,112,128],  stride: [1,112,37632]    <----- transpose of combo features
Exception in thread "main" java.lang.IllegalStateException: Rank is 3 rows() call is not valid
	at org.nd4j.linalg.api.ndarray.BaseNDArray.rows(BaseNDArray.java:4037)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.mmul(BaseNDArray.java:2808)
	at edu.southwestern.util.graphics.NeuralStyleTransfer.derivativeLossStyleInLayer(NeuralStyleTransfer.java:194)
	at edu.southwestern.util.graphics.NeuralStyleTransfer.main(NeuralStyleTransfer.java:306)


From looking at the shapes, I understand why the matrix operation does not work. However, I don't know why the sizes are wrong for the given layer. I'll need to look harder at the original algorithm to figure out what the shapes should be at each stage.

**Comment by klevis** [comment]
Hi ,
I have worked on the above problem and it seems now the shapes are working but there is a problem with exploding pixels values. Back-propagation produces really big values and is not getting any better in time. 
https://github.com/klevis/MM-NEAT/blob/dev/src/main/java/edu/southwestern/util/graphics/NeuralStyleTransfer.java
Any idea from here @schrum2 , @AlexDBlack ?

**Comment by schrum2** [comment]
I'm excited to here that you made progress with this, but I had actually given up. Or rather, I found a work around, in that I wrote some Java code that launches a Python process which executes the Neural Style Transfer Algorithm. I had to modify the Python code to read json representations of images from stdin and then output the resulting image as a json value that Java could read back in and display.

The code is currently in the dev branch of MM-NEAT at:
https://github.com/schrum2/MM-NEAT/blob/dev/src/main/java/edu/southwestern/util/graphics/PythonNeuralStyleTransfer.java
The Python code is in:
https://github.com/schrum2/MM-NEAT/tree/dev/src/main/python
and comes from
https://github.com/anishathalye/neural-style
You will need to follow their install instructions, and also download the VGG net weights that they provide a link to. If you can get the Python program to work, then feel free to play with the application in:
https://github.com/schrum2/MM-NEAT/blob/dev/src/main/java/edu/southwestern/tasks/interactive/remixbreeder/PictureStyleBreederTask.java
Just launch the main method. It's still a prototype, but is kind of fun to mess around with, assuming you can make the Python program behave on your system.


Still, I would prefer a Java-only solution, since that would be much more portable than this one, which involves many steps, and interactions between different systems that could be configured wrong. I'm happy to hear about any progress/code for a Java version of Neural Style Transfer.

**Comment by agibsonccc** [comment]
@schrum2 if you can't reliably build it with dl4j by yourself maybe wait till our model import mechanism with nd4j is done? You could try it then I suppose.

**Comment by schrum2** [comment]
I'm not sure how that would help. The problem here isn't the model, but the algorithm, isn't it? Unless you are implying that your model import mechanism for ND4J will provide me with a means to automatically calculate the gradients for some arbitrary loss function, as is done in the tensorflow implementation of this algorithm.

If you are planning on adding this functionality to DL4J (which is of course a good idea, since that's what all the other deep learning libraries do), then might I propose the neural style transfer algorithm as a test case? It is a fairly popular example of needing to compute gradients, but without actually using them to update network parameters.

**Comment by agibsonccc** [comment]
Yes that's exactly what I'm implying. We have autodiff and everything right in there for you. So you can import a pre existing model *or* you can just use autodiff. Either will work.

**Comment by schrum2** [comment]
Awesome! I look forward to it, thought I'll definitely be waiting until it is available on Maven.

**Comment by saudet** [comment]
@schrum2 It is available as SNAPSHOT artifacts: https://deeplearning4j.org/snapshots

**Comment by klevis** [comment]
HI @schrum2  In case you are still interested , I was able to get some result with this:
Style : https://github.com/klevis/MM-NEAT/blob/dev/data/imagematch/style2.jpg
Content: https://github.com/klevis/MM-NEAT/blob/dev/data/imagematch/content2.jpg
Combined Image: https://github.com/klevis/MM-NEAT/blob/dev/data/iteration1000_old.jpg

For now only  works well with one style layer. It becomes a bit unstable with more layers as cost is going slowly up regardless of regularization parameter decreased.

Maybe something wrong with partial derivatives with multiple layers... 
For now I do a back-propagation per each layers separately , do you if this is the way to go?
https://github.com/klevis/MM-NEAT/blob/dev/src/main/java/edu/southwestern/util/graphics/NeuralStyleTransfer.java

**Comment by klevis** [comment]
It looks like the problem was rather with the cost function showing wrong progress values. So now it works fine also with multiple layers and showing better results. The only downside now is the fact that is extremely slow...
Is there any example or suggestion how to apply ADAM separately? -maybe it helps

**Comment by agibsonccc** [comment]
@klevis we would love to get this in to the examples. Mind sending a PR? Great job!

**Comment by AlexDBlack** [comment]
Yeah, this is cool. +1 on the example PR request.

As to slowness: the code doesn't look too bad - though there's a few minor optimizations I can see (in-place vs. copy ops). Would need to run it in profiler/debugger.

**Comment by klevis** [comment]
I will clean up the code a bit and send a PR soon, to the dl4j-examples right?

**Comment by agibsonccc** [comment]
That or model zoo. Thoughts @crockpotveggies ?

**Comment by schrum2** [comment]
@klevis Would you mind doing a PR for the MM-NEAT project too (to dev branch)? I was about to copy over your code, but would rather maintain a record of your contribution.

**Comment by klevis** [comment]
Hi 
I sent the PR request: https://github.com/deeplearning4j/dl4j-examples/pull/639
I think running on GPU it will bring better results as you can leave it run for more iterations. 
For now images are only generated with 300-380 iterations.

**Comment by AlexDBlack** [comment]
Closing this as no longer relevant, given @klevis 's contribution to the examples :)
